### PR TITLE
refactor: use Formik on Earn screen

### PR DESCRIPTION
**Pull request by theborakompanioni**
Uses Formik for form validation on Earn screen.

Further adaptions: 
- Don't persist form values to local storage on every update, but just when the form is actually submitted
- Separate info and error alert - those are not mutually exclusive

### CODE REVIEW

File: src/components/Earn.jsx [hunk 1]
```diff
@@ -290,135 +317,145 @@ export default function Earn() {
...
+            <Formik initialValues={initialValues} validate={validate} onSubmit={onSubmit}>
+              {({ handleSubmit, setFieldValue, handleChange, handleBlur, values, touched, errors, isSubmitting }) => (
+                <rb.Form onSubmit={handleSubmit} noValidate>
+                  {!serviceInfo?.makerRunning && !isWaitingMakerStart && !isWaitingMakerStop && (
+                    <>
+                      {settings.useAdvancedWalletMode && (
+                        <rb.Form.Group className="mb-3" controlId="offertype">
+                          <ToggleSwitch
+                            label={t('earn.toggle_rel_offer')}
+                            initialValue={values.offertype === OFFERTYPE_REL}
```
Comment (dnlggr) [inline src/components/Earn.jsx:332]: Would something like `initialValue={values.offertype === initialValues.offertype}` here?

**Review by dnlggr — COMMENTED** [review]
Nice! 👏 👌  Left a couple of minor style-related comments inline, let me know what you think!

Something I notices was that -- while everything worked as expected -- the feedback for an invalid absolute fee was shown next to the wrong field (the amount field instead of the fee field). For the relative fee it was correct though. See screenshots below.

Screenshot 2022-04-28 at 18 05 23
Screenshot 2022-04-28 at 18 05 40

File: src/components/Earn.jsx [hunk 2]
```diff
@@ -80,70 +82,62 @@ const YieldgenReport = ({ lines, maxAmountOfRows = 25 }) => {
...
+  window.localStorage.setItem('jm-feeRel', value)
+}
+
+const persistFeeAbs = (value) => {
+  window.localStorage.setItem('jm-feeAbs', value)
+}
+
+const persistMinsize = (value) => {
+  window.localStorage.setItem('jm-minsize', value)
+}
```
[the remark this answers is not in the document]
Comment (theborakompanioni) [reply src/components/Earn.jsx:113]: Done.

File: src/components/Earn.jsx [hunk 3]
```diff
@@ -276,6 +245,64 @@ export default function Earn() {
...
+    minsize: parseInt(window.localStorage.getItem('jm-minsize'), 10) || 100_000,
+  }
+
+  const validate = (values) => {
+    const errors = {}
+    if (!OFFERTYPES.includes(values.offertype)) {
+      // currently no need for translation, this should never occur -> input is controlled by toggle
+      errors.offertype = `Offertype must be one of ${OFFERTYPES.join(',')}`
+    }
+    if (!(values.feeRel >= feeRelMin && values.feeRel <= feeRelMax)) {
```
[the remark this answers is not in the document]
Comment (theborakompanioni) [reply src/components/Earn.jsx:265]: Done.

File: src/components/Earn.jsx (hunk 1 above)
Comment (theborakompanioni) [reply src/components/Earn.jsx:332]: Hmm.. I don't know if it'd be beneficial as the `onToggle` handler would also have to be adapted if the value of `initialValues.offertype` changed. The toggle is tailored towards being active if `offertype === OFFERTYPE_REL`. Does that make sense?

File: src/components/Earn.jsx [hunk 4]
```diff
@@ -276,6 +245,64 @@ export default function Earn() {
...
 
+  const feeRelMin = 0.0
+  const feeRelMax = 0.1 // 10%
+  const feeRelPercentageStep = 0.0001
+
+  const initialValues = {
+    offertype: (settings.useAdvancedWalletMode && window.localStorage.getItem('jm-offertype')) || OFFERTYPE_REL,
+    feeRel: parseFloat(window.localStorage.getItem('jm-feeRel')) || 0.000_3,
+    feeAbs: parseInt(window.localStorage.getItem('jm-feeAbs'), 10) || 250,
+    minsize: parseInt(window.localStorage.getItem('jm-minsize'), 10) || 100_000,
```
[the remark this answers is not in the document]
Comment (theborakompanioni) [reply src/components/Earn.jsx]: Done.

**Comment by dnlggr** [comment]
Nice feedback is now shown correctly everytime! 👌 

I have one more edge case: When you enter an invalid absolute fee and then switch to relative fees, whatever you enter there will remain invalid. Also works the other way around (invalid relative fee -> switch to absolute fees -> everything will be invalid). Somehow the invalid value seems to get carried over when you switch fee types and then remain there. CAn you reproduce this?

https://user-images.githubusercontent.com/10026790/165915093-d0c5be95-6d17-4f4d-9ef2-d41b573f4d69.mov

**Comment by theborakompanioni** [comment]
> Can you reproduce this?

Yes.

> When you enter an invalid absolute fee and then switch to relative fees, whatever you enter there will remain invalid. 

In the back of my head I knew something like this was possible but why exactly didn't I test this behaviour? I don't know..

One option would be to completely separate the forms for relative and absolute offers.. However, as all values are sent in the request body, I think the expected behaviour is to only validate the relevant fields and subsequently also only store these fields in local storage. Does this sound reasonable to you or would you split the forms?

**Comment by dnlggr** [comment]
> all values are sent in the request body

Oh yeah just saw that. That's a bit strange. Are we fine with sending invalid values for e.g. the absolute fee when we know that the fee type is relative anyway? 🤔 I'd prefer only the relevant fee to be sent but I guess that's out of our hands at this point. 😅 What do you think?

**Comment by theborakompanioni** [comment]
> > all values are sent in the request body
> 
> Oh yeah just saw that. That's a bit strange. Are we fine with sending invalid values for e.g. the absolute fee when we know that the fee type is relative anyway? thinking I'd prefer only the relevant fee to be sent but I guess that's out of our hands at this point. sweat_smile What do you think?

It's funny because all values must be present in the request, but they are accepted even when they are invalid.. current changes sets the ignored value to zero (thought about setting it to a maximum value, but what is the max for an absolute fee?). Goal is to mitigate the need to ship hotfixes in case they will be validated in the future. We do not have tests against an actual api implementation and we can only control the version of our docker images, so we should expect some users to also run different server versions – but of course we cannot control possible serverside breaking changes, just mitigate them.

I am aware that the current change is not optimal, as it has conditionals based on `offertype` in multiple places – should be way better with typescript and distinct types, but I resisted to divert from original goal, which is using Formik consistently. I hope that is acceptable.

**Review by dnlggr — APPROVED** [review]
Yeah I think this is an acceptable solution now. It's a bit unfortunate how the backend works here but we have to live with it for now I guess. ✅ LGMT! Good job!

**Pull request merged**